File: Visual_Basic/Trees/BinaryTree/BinaryTree.vb
'******************************************************
' *  BinaryTree.vb
' *  Created by Stephen Hall on 9/25/17.
' *  Copyright (c) 2016 Stephen Hall. All rights reserved.
' *  A Binary Tree implementation in Visual Basic
' *******************************************************

Namespace Trees.BinaryTree
	''' <summary>
	''' Binary Tree Class
	''' </summary>
	''' <typeparam name="T">Generic Type</typeparam>
	Public Class BinaryTree(Of T As IComparable)
		''' <summary>
		''' Node class for the binary tree
		''' </summary>
		Public Class Node
			''' <summary>
			''' Public Properties for  the Node class
			''' </summary>
			Public Property Data As T
			Public Property Left As Node
            Public Property Right As Node
			Public Property Parent As Node

			''' <summary>
			''' Node class constructor             
			''' </summary>
			''' <param name="data1">Data to be held in the Node</param>
			Public Sub New(data1 As T)
				Left = InlineAssignHelper(Right, Nothing)
				Data = data1
			End Sub
			Private Shared Function InlineAssignHelper(Of T)(ByRef target As T, value As T) As T
				target = value
				Return value
			End Function
		End Class

		''' <summary>
		''' Private Binary Tree members
		''' </summary>
		Private _root As Node

		''' <summary>
		''' Binary tree class constructor
		''' </summary>
		Public Sub New()
			_root = Nothing
		End Sub

		''' <summary>
		''' Compares the data in given node to the give data for equality
		''' </summary>
		''' <param name="data">data to compare</param>
		''' <param name="node">Node to compare to</param>
		''' <returns>1 if greater then, 0 if equal to, -1 if less then</returns>
		Private Shared Function Compare(data As T, node As Node) As Integer
			Return node.Data.CompareTo(data)
		End Function

		''' <summary>
		''' Inserts a new node into the tree
		''' </summary>
		''' <param name="data">data to insert into the tree</param>
		''' <returns>Node inserted into the tree</returns>
		Public Function Insert(data As T) As Node
			Return If(_root Is Nothing, InlineAssignHelper(_root, New Node(data)), InsertHelper(data, _root))
		End Function

		''' <summary>
		''' Recursive helper function to inset a new node into the tree
		''' </summary>
		''' <param name="data">Data to insert into the tree</param>
		''' <param name="node">urrent node in the tree</param>
		''' <returns>Node inserted into the tree</returns>
		Private Shared Function InsertHelper(data As T, node As Node) As Node
			Try
				Dim cmp As Integer = Compare(data, node)
				Select Case cmp
					Case 1
						If node.Right Is Nothing Then
							Dim newNode As New Node(data)
							node.Right = newNode
							newNode.Parent = node
							Return newNode
						Else
							Return InsertHelper(data, node.Right)
						End If
					Case Else

						If node.Left Is Nothing Then
							Dim newNode As New Node(data)
							node.Left = newNode
							newNode.Parent = node
							Return newNode
						Else
							Return InsertHelper(data, node.Left)
						End If
				End Select
			Catch generatedExceptionName As Exception
				Return Nothing
			End Try
		End Function

		''' <summary>
		''' Removes a Node from the tree
		''' </summary>
		''' <param name="data">Data to remove from the tree</param>
		''' <returns>Node removed from the tree</returns>
		Public Function Remove(data As T) As Node
			Return If(_root IsNot Nothing, Nothing, RemoveHelper(data, _root))
		End Function

		''' <summary>
		''' Recursive helper function to remove a node from the tree
		''' </summary>
		''' <param name="data">Data to remove</param>
		''' <param name="node">urrent node</param>
		''' <returns>Node removed from the tree</returns>
		Private Function RemoveHelper(data As T, node As Node) As Node
			Try
				Dim cmp As Integer = Compare(data, node)
				If cmp = 1 Then
					Return RemoveHelper(data, node.Right)
				End If
				If cmp = -1 Then
					Return RemoveHelper(data, node.Left)
				End If
				If cmp = 0 Then
					'has no children
					Dim tempNode As Node

					If node.Left Is Nothing AndAlso node.Right Is Nothing Then
						node = node.Parent
					End If

					'has one child
					If node.Parent IsNot Nothing Then
						tempNode = node.Parent
					Else
						' this is the root node
						If node.Right IsNot Nothing Then
							tempNode = node.Right
							tempNode.Parent = Nothing
							_root.Right = Nothing
							_root = tempNode
							Return node
						End If
						tempNode = node.Left
						If tempNode IsNot Nothing Then
							tempNode.Parent = Nothing
							_root.Left = Nothing
							_root = tempNode
						End If
						Return node
					End If
					If tempNode.Left Is node Then
						If node.Left IsNot Nothing Then
							tempNode.Left = node.Left
							node.Left.Parent = tempNode
							Return node
						End If
						If node.Right IsNot Nothing Then
							tempNode.Right = node.Right
							node.Right.Parent = tempNode
							Return node
						End If
					End If
					If tempNode.Right Is node Then
						If node.Left IsNot Nothing Then
							tempNode.Left = node.Left
							node.Left.Parent = tempNode
							Return node
						End If
						If node.Right IsNot Nothing Then
							tempNode.Right = node.Right
							node.Right.Parent = tempNode
							Return node
						End If
					ElseIf node.Left IsNot Nothing AndAlso node.Right IsNot Nothing Then
						' test for 2 children
						tempNode = node.Right
						' find min value of right child tree to replace the node
						While tempNode.Left IsNot Nothing
							tempNode = tempNode.Left
						End While
						Dim temp As T = node.Data
						node.Data = tempNode.Data
						' replace the node with the tempNode
						tempNode.Data = temp
						tempNode.Parent.Left = Nothing
						tempNode.Parent = Nothing
						Return tempNode
					End If
				End If
			Catch generatedExceptionName As Exception
				Return Nothing
			End Try
			Return Nothing
		End Function

		''' <summary>
		''' Returns the smallest node in the tree
		''' </summary>
		''' <returns>Smallest Node int he tree</returns>
		Public Function GetMin() As Node
			If _root Is Nothing Then
				Return Nothing
			End If

			Dim node As Node = _root

			While node.Left IsNot Nothing
				node = node.Left
			End While

			Return node
		End Function

		''' <summary>
		''' Return the largest node in the tree
		''' </summary>
		''' <returns>argest Node in the tree</returns>
		Public Function GetMax() As Node
			If _root Is Nothing Then
				Return Nothing
			End If

			Dim node As Node = _root

			While node.Right IsNot Nothing
				node = node.Right
			End While

			Return node
		End Function

		''' <summary>
		''' Prints out the tree using Pre Order Traversal
		''' </summary>
		''' <param name="node">Node to start the Pre Order Traversal at</param>
		Public Sub PreOrederTraversal(node As Node)
			If node IsNot Nothing Then
				Console.WriteLine(node.Data)
				PreOrederTraversal(node.Left)
				PreOrederTraversal(node.Right)
			End If
		End Sub

		''' <summary>
		''' Prints out the Tree using Post Order Traversal
		''' </summary>
		''' <param name="node">Node to start the Post Order Traversal at</param>
		Public Sub PostPrderTraversal(node As Node)
			If node IsNot Nothing Then
				PostPrderTraversal(node.Left)
				PostPrderTraversal(node.Right)
				Console.WriteLine(node.Data)
			End If
		End Sub

		''' <summary>
		''' Pints out the tree using in order Traversal
		''' </summary>
		''' <param name="node">Node to start the In Order Traversal at</param>
		Public Sub InOrderedTraversal(node As Node)
			If node IsNot Nothing Then
				InOrderedTraversal(node.Left)
				Console.WriteLine(node.Data)
				InOrderedTraversal(node.Right)
			End If
		End Sub

		''' <summary>
		''' Prints out the tree using Depth First Search
		''' </summary>
		''' <param name="node">Node to start the Depth First Search at</param>
		Public Sub DepthFirstSearch(node As Node)
			If node Is Nothing Then
				Return
			End If

			Dim stack As New Stack(Of Node)()
			stack.Push(node)

			While stack.Count > 0
				Console.WriteLine(InlineAssignHelper(node, stack.Pop()))
				If node.Right IsNot Nothing Then
					stack.Push(node.Right)
				End If
				If node.Left IsNot Nothing Then
					stack.Push(node.Left)
				End If
			End While
		End Sub

		''' <summary>
		''' Prints out the tree using breadth first search
		''' </summary>
		''' <param name="node">Node to start Breadth First Search at</param>
		Public Sub BreadthFirstSearch(node As Node)
			If node Is Nothing Then
				Return
			End If

			Dim queue As New Queue(Of Node)()
			queue.Enqueue(node)

			While queue.Count > 0
				Console.WriteLine(InlineAssignHelper(node, queue.Dequeue()))
				If node.Left IsNot Nothing Then
					queue.Enqueue(node.Left)
				End If
				If node.Right IsNot Nothing Then
					queue.Enqueue(node.Right)
				End If
			End While
		End Sub
		Private Shared Function InlineAssignHelper(Of T)(ByRef target As T, value As T) As T
			target = value
			Return value
		End Function
	End Class
End Namespace

File: PHP/Trees/BinaryTree.php
<?php namespace datastructures\binarytree;
/*******************************************************
 *  BinaryTree.php
 *  Created by Stephen Hall on 11/01/17.
 *  Copyright (c) 2016 Stephen Hall. All rights reserved.
 *  A Binary Tree implementation in PHP
 ********************************************************/

/**
 * Class Node
 */
class Node {
    /**
     * @var $left : left leaf in the tree
     * @var $right : right leaf in the tree
     * @var $parent : parent node in the tree
     * @var $data : data to be help by the node
     */
    public $left;
    public $right;
    public $parent;
    public $data;

    /**
     * Node constructor.
     * @param $data: data to be help in the node
     */
    public function __construct($data) {
        $this->left = $this->right = null;
        $this->data = $data;
    }
}

/**
 * Class BinaryTree
 */
class BinaryTree {

    /**
     * @var $root : root node of the tree
     */
    private $root;

    /**
     * BinaryTree constructor.
     */
    public function __construct() {
        $this->root = null;
    }

    /**
     * Compares the data in given node to the give data for equality
     * @param $data : data to compare
     * @param $node : Node to compare to
     * @return int : 1 if greater then, 0 if equal to, -1 if less then
     */
    private function compare($data, $node) {
        return (strcmp((string)$node->data, (string)$data));
    }

    /**
     * Inserts a new node into the tree
     * @param $data : data to insert into the tree
     * @return Node : Node inserted into the tree
     */
    public function Insert($data) {
        if($this->root == null) {
            return ($this->root = new Node($data));
        }
        return $this->insertHelper($data, $this->root);
    }

    /**
     * Recursive helper function to inset a new node into the tree
     * @param $data : Data to insert into the tree
     * @param $node : Current node in the tree
     * @return Node : Node inserted into the tree or null
     */
    private function insertHelper($data, $node) {
        $cmp = $this->compare($data, $node);

        switch($cmp) {
            case 1:
                if ($node->right == null) {
                    $newNode = new Node($data);
                    $node->right = $newNode;
                    $newNode->parent = $node;
                    return $newNode;
                }
                else
                    return $this->insertHelper($data, $node->right);

            default:
                if ($node->left == null) {
                    $newNode = new Node($data);
                    $node->left = $newNode;
                    $newNode->parent = $node;
                    return $newNode;
                }
                else
                    return $this->insertHelper($data, $node->left);
        }
    }

    /**
     * Removes a Node from the tree
     * @param $data : Data to remove from the tree
     * @return Node : Node removed from the tree
     */
    public function Remove($data) {
        if($this->root != null)
            return null;
        return $this->removeHelper($data, $this->root);
    }

    /**
     * Recursive helper function to remove a node from the tree
     * @param $data : Data to remove
     * @param $node : Current node in the tree
     * @return Node : Node removed from the tree or null
     */
    private function removeHelper($data, $node) {
        $cmp = $this->compare($data, $node);

        if($cmp == 1)
            return $this->removeHelper($data, $node->right);

        if($cmp == -1)
            return $this->removeHelper($data, $node->left);

        if($cmp == 0) {
            //has no children
            if($node->left == null && $node->right == null) {
                $node = $node->parent;
            }
            $tempNode = null;
            //has one child
            if($node->parent != null) {
                $tempNode = $node->parent;
            } else {   // this is the root node
                if($node->right != null) {
                    $tempNode = $node->right;
                    $tempNode->parent = null;
                    $this->root->right = null;
                    $this->root = $tempNode;
                    return $node;
                } else {
                    $tempNode = $node->left;
                    $tempNode->parent = null;
                    $this->root->left = null;
                    $this->root = $tempNode;
                    return $node;
                }
            }
            if($tempNode->left == $node) {
                if($node->left != null) {
                    $tempNode->left = $node->left;
                    $node->left->parent = $tempNode;
                    return $node;
                }
                else if($node->right != null) {
                    $tempNode->right = $node->right;
                    $node->right->parent = $tempNode;
                    return $node;
                }
            }
            if($tempNode->right == $node) {
                if($node->left != null) {
                    $tempNode->left = $node->left;
                    $node->left->parent = $tempNode;
                    return $node;
                }
                else if($node->right != null) {
                    $tempNode->right = $node->right;
                    $node->right->parent = $tempNode;
                    return $node;
                }
            }
            else if ($node->left != null && $node->right != null) {
                // test for 2 children
                $tempNode = $node->right;

                // find min value of right child tree to replace the node
                while($tempNode->left != null)
                    $tempNode = $tempNode->left;

                $temp = $node->data;
                $node->data = $tempNode->data;
                // replace the node with the tempNode
                $tempNode->data = $temp;
                $tempNode->parent->left = null;
                $tempNode->parent = null;
                return $tempNode;
            }
        }
    }

    /**
     * Returns the smallest node in the tree
     * @return Node : Smallest Node in the tree
     */
    public function GetMin() {
        if($this->root == null)
            return null;

        $node = $this->root;

        while($node->left != null)
            $node = $node->left;

        return $node;
    }

    /**
     * Returns the largest node in the tree
     * @return Node|null: Largest Node in the tree
     */
    public function GetMax() {
        if($this->root == null)
            return null;

        $node = $this->root;

        while($node->right != null)
            $node = $node->right;

        return $node;
    }

    /**
     * Prints out the tree from the given node using Pre Order Traversal
     * @param $node : Node to start the Pre Order Traversal at
     */
    public function PreOrderTraversal($node) {
        if($node != null){
            echo $node->data." ";
            $this->PreOrderTraversal($node->left);
            $this->PreOrderTraversal($node->right);
        }
    }

    /**
     * Prints out the Tree from the given node using Post Order Traversal
     * @param $node : Node to start the Post Order Traversal at
     */
    public function PostOrderTraversal($node) {
        if($node != null) {
            $this->PostOrderTraversal($node->left);
            $this->PostOrderTraversal($node->right);
            echo $node->data." ";
        }
    }

    /**
     * Pints out the tree from the given node using in order Traversal
     * @param $node: Node to start the In Order Traversal at
     */
    public function InOrderedTraversal($node) {
        if($node != null) {
            $this->InOrderedTraversal($node->left);
            echo $node->data." ";
            $this->InOrderedTraversal($node->right);
        }
    }

    /**
     * Prints out the tree from the given node using Depth First Search
     * @param $node: Node to start the Depth First Search at
     */
    public function DepthFirstSearch($node){
        if($node == null)
            return;

        $stack = [];
        array_push($stack,$node);

        while (!empty($stack)) {
            $node = array_pop($stack);
            echo $node->data." ";

            if($node->right != null)
                array_push($stack, $node->right);

            if($node->left != null)
                array_push($stack, $node->left);
        }
    }

    /**
     * Prints out the tree from the given node using breadth first search
     * @param $node : Node to start Breadth First Search at
     */
    public function BreadthFirstSearch($node){
        if($node == null)
            return;

        $queue = [];
        array_push($queue,$node);

        while (count($queue) > 0) {
            $node = array_shift($queue);
            echo $node." ";

            if($node->left != null)
                array_push($queue,$node->left);

            if($node->right != null)
                array_push($queue,$node->right);
        }
    }
}